File: day03/ring.c
#include <stdio.h>
#include <string.h>
// Absolute function
int abs(int x) { return (x < 0) ? -x : x; }

// Min function
int min(int x, int y) { return (x < y) ? x : y; }

int min_steps_loop(char *ring, char *key) {
  int nk = strlen(key);
  int nr = strlen(ring);

  int distance[100][100];

  int indices[26][100] = {[0 ... 25][0 ... 99] = -1};
  for (int i = 0; i < 26; i++) {
    int index = 0;
    int ch = 'a' + i;
    for (int j = 0; j < nr; j++) {
      if (ring[j] == ch) {
        indices[i][index] = j;
        index++;
      }
    }
  }

  for (int key_index = nk - 1; key_index > 0; key_index--) {
    int x = 0;
    while (indices[key[key_index - 1] - 'a'][x] != -1) {
      int ring_index = indices[key[key_index - 1] - 'a'][x];
      int res = 1000000;

      int y = 0;
      while (indices[key[key_index] - 'a'][y] != -1) {
        int char_index = indices[key[key_index] - 'a'][y];
        int dist = abs(ring_index - char_index);
        dist = min(dist, nr - dist);

        int foo =
            1 + dist +
            ((key_index == nk - 1) ? 0 : distance[char_index][key_index + 1]);

        res = min(res, foo);
        y++;
      }
      distance[ring_index][key_index] = res;

      x++;
    }
  }

  int key_index = 0;
  int ring_index = 0;
  int res = 1000000;

  int y = 0;
  while (indices[key[key_index] - 'a'][y] != -1) {
    int char_index = indices[key[key_index] - 'a'][y];
    int dist = abs(ring_index - char_index);
    dist = min(dist, nr - dist);

    int foo = 1 + dist +
              ((key_index == nk - 1) ? 0 : distance[char_index][key_index + 1]);

    res = min(res, foo);
    y++;
  }
  distance[0][0] = res;
  return distance[0][0];
}

int main() {
  for (int i = 0; i < 100; i++) {
    min_steps_loop("godding", "gd");
    min_steps_loop("godding", "godding");
    min_steps_loop("abcde", "ade");
    min_steps_loop("y", "y");
    min_steps_loop("dccjgocjnscdalkwrlzmsrhnprvvsvqjtqnzirtaasdeldiiokttozjf"

                   "kwjghhaibtzkoepdvfkhxgmxwtwsrgiryzqljpsntjei",
                   "lsdsvjtj");
    min_steps_loop("zmgddxyivijgullayjbxlmcbirsgpaeqemefpshtehczznvqjaetauso"

                   "agpkqjtewszqihnzevhbcgpmmtdbyesgxjulyzshnksc",
                   "yxebsleaklhkjlx");
    min_steps_loop("gyzhbpwtmixjdmozuctjctsvcdxtpbtaftuqqirabgmnibnanegqxivx"

                   "ljloodnmmylisompbtccwksvxcdxoixgjibhkmsujdvx",
                   "tihsvtysmktbjtjkcvirxbixbx");
    min_steps_loop("cimrihxbkmotdfudkkrovdydtyamiqdsogsaffnvdqucwfdrtrppmsjs"

                   "mbekuxzltpktpzcuayjtowoyemlzggktxzkqzosenqgb",
                   "dnkvlggcpqslgsbmnyiukfcglsgglvbqvoosetoktrtgezvbukxqsmie"

                   "foffbsgdoomhgfmqumsmmdspqckblzcawzupwagsmwii");
    min_steps_loop("wmgzcptgkwwnezrymhqtxlpbavqpioymjqifbdiwtldfqgcycqlfjxlm"

                   "yeqtynjpxprzqelgecdqlrbayhmbvneyfoopwlhdkvzc",
                   "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"

                   "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa");
    min_steps_loop("dslgzxnzgjncgeofucxncbjwaehabidlohnklkcgsxugshlhgjsiqjwf"

                   "xeignrgcfrshxnecfssmqxbcxdckfpodqazgoddkbmsh",
                   "fxseifxczzcnhgscdgidkcqshfqbeigceaccszjzxgjghkjdkskjocoh"

                   "phmsdxogxacklrcxcuaxdkscnkdkjajggqxhqkaumgcf");
    min_steps_loop("qdwcejgziqiulnvnttdkywpbpywduevimnutnzlvfefsntqgdvfrqnfo"

                   "knvholqozccyhmdjbqizbuvbfafybfuqfdrosqlztazl",
                   "vteilniztwdsoaunpzhztvdddelkndvrwlwctinneujuqkvkvgeeldlk"

                   "umnqbdzmnklgfortizljnrveczpynqyqqgffyqqunuqp");
  }

  printf("%d\n", min_steps_loop("godding", "gd"));
  printf("%d\n", min_steps_loop("godding", "godding"));
  printf("%d\n", min_steps_loop("abcde", "ade"));
  printf("%d\n", min_steps_loop("y", "y"));
  printf("%d\n", min_steps_loop("eh", "h"));
  printf("%d\n",
         min_steps_loop("dccjgocjnscdalkwrlzmsrhnprvvsvqjtqnzirtaasdeldiiokt"

                        "tozjfkwjghhaibtzkoepdvfkhxgmxwtwsrgiryzqljpsntjei",

                        "lsdsvjtj"));
  printf("%d\n",
         min_steps_loop("zmgddxyivijgullayjbxlmcbirsgpaeqemefpshtehczznvqjae"

                        "tausoagpkqjtewszqihnzevhbcgpmmtdbyesgxjulyzshnksc",

                        "yxebsleaklhkjlx"));
  printf("%d\n",
         min_steps_loop("gyzhbpwtmixjdmozuctjctsvcdxtpbtaftuqqirabgmnibnaneg"

                        "qxivxljloodnmmylisompbtccwksvxcdxoixgjibhkmsujdvx",

                        "tihsvtysmktbjtjkcvirxbixbx"));
  printf("%d\n",
         min_steps_loop(
             "cimrihxbkmotdfudkkrovdydtyamiqdsogsaffnvdqucwfdrtrppmsjsmbe"

             "kuxzltpktpzcuayjtowoyemlzggktxzkqzosenqgb",
             "dnkvlggcpqslgsbmnyiukfcglsgglvbqvoosetoktrtgezvbukxqsmiefof"

             "fbsgdoomhgfmqumsmmdspqckblzcawzupwagsmwii"));
  printf("%d\n",
         min_steps_loop(
             "wmgzcptgkwwnezrymhqtxlpbavqpioymjqifbdiwtldfqgcycqlfjxlmyeq"

             "tynjpxprzqelgecdqlrbayhmbvneyfoopwlhdkvzc",
             "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"

             "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"));
  printf("%d\n",
         min_steps_loop(
             "dslgzxnzgjncgeofucxncbjwaehabidlohnklkcgsxugshlhgjsiqjwfxei"

             "gnrgcfrshxnecfssmqxbcxdckfpodqazgoddkbmsh",
             "fxseifxczzcnhgscdgidkcqshfqbeigceaccszjzxgjghkjdkskjocohphm"

             "sdxogxacklrcxcuaxdkscnkdkjajggqxhqkaumgcf"));
  printf("%d\n",
         min_steps_loop(
             "qdwcejgziqiulnvnttdkywpbpywduevimnutnzlvfefsntqgdvfrqnfoknv"

             "holqozccyhmdjbqizbuvbfafybfuqfdrosqlztazl",
             "vteilniztwdsoaunpzhztvdddelkndvrwlwctinneujuqkvkvgeeldlkumn"

             "qbdzmnklgfortizljnrveczpynqyqqgffyqqunuqp"));
  return 0;
}

File: day03/ring.rs
// https://leetcode.com/problems/freedom-trail/description/

struct Solution;
impl Solution {
    pub fn min_steps(
        ringIndex: usize,
        keyIndex: usize,
        ring: &String,
        key: &[u8],
        distance: &mut Vec<Vec<i32>>,
    ) -> i32 {
        if distance[ringIndex][keyIndex] != -1 {
            return distance[ringIndex][keyIndex];
        }
        let res = ring
            .chars()
            .enumerate()
            .filter(|(_, x)| *x == key[0] as char)
            .map(|(i, _)| {
                let mut dist = ringIndex as i32 - i as i32;
                dist = dist.abs();
                dist = dist.min(ring.len() as i32 - dist);
                (i, dist)
            })
            .map(|(i, x)| {
                x + 1
                    + if key.len() == 1 {
                        0
                    } else {
                        Self::min_steps(i, keyIndex + 1, &ring, &key[1..], distance)
                    }
            })
            .min()
            .unwrap();
        distance[ringIndex][keyIndex] = res;
        res
    }

    pub fn find_rotate_steps(ring: String, key: String) -> i32 {
        let mut distance = vec![vec![-1; key.len()]; ring.len()];
        Self::min_steps(0, 0, &ring, &key.as_bytes(), &mut distance)
    }
}

fn main() {
    for _ in 0..100 {
        Solution::find_rotate_steps("godding".to_string(), "gd".to_string());
        Solution::find_rotate_steps("godding".to_string(), "godding".to_string());
        Solution::find_rotate_steps("abcde".to_string(), "ade".to_string());
        Solution::find_rotate_steps("y".to_string(), "y".to_string());
        Solution::find_rotate_steps("dccjgocjnscdalkwrlzmsrhnprvvsvqjtqnzirtaasdeldiiokttozjfkwjghhaibtzkoepdvfkhxgmxwtwsrgiryzqljpsntjei".to_string(), "lsdsvjtj".to_string());
        Solution::find_rotate_steps("zmgddxyivijgullayjbxlmcbirsgpaeqemefpshtehczznvqjaetausoagpkqjtewszqihnzevhbcgpmmtdbyesgxjulyzshnksc".to_string(), "yxebsleaklhkjlx".to_string());
        Solution::find_rotate_steps("gyzhbpwtmixjdmozuctjctsvcdxtpbtaftuqqirabgmnibnanegqxivxljloodnmmylisompbtccwksvxcdxoixgjibhkmsujdvx".to_string(), "tihsvtysmktbjtjkcvirxbixbx".to_string());
        Solution::find_rotate_steps("cimrihxbkmotdfudkkrovdydtyamiqdsogsaffnvdqucwfdrtrppmsjsmbekuxzltpktpzcuayjtowoyemlzggktxzkqzosenqgb".to_string(), "dnkvlggcpqslgsbmnyiukfcglsgglvbqvoosetoktrtgezvbukxqsmiefoffbsgdoomhgfmqumsmmdspqckblzcawzupwagsmwii".to_string());
        Solution::find_rotate_steps("wmgzcptgkwwnezrymhqtxlpbavqpioymjqifbdiwtldfqgcycqlfjxlmyeqtynjpxprzqelgecdqlrbayhmbvneyfoopwlhdkvzc".to_string(), "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa".to_string());
        Solution::find_rotate_steps("dslgzxnzgjncgeofucxncbjwaehabidlohnklkcgsxugshlhgjsiqjwfxeignrgcfrshxnecfssmqxbcxdckfpodqazgoddkbmsh".to_string(), "fxseifxczzcnhgscdgidkcqshfqbeigceaccszjzxgjghkjdkskjocohphmsdxogxacklrcxcuaxdkscnkdkjajggqxhqkaumgcf".to_string());
        Solution::find_rotate_steps("qdwcejgziqiulnvnttdkywpbpywduevimnutnzlvfefsntqgdvfrqnfoknvholqozccyhmdjbqizbuvbfafybfuqfdrosqlztazl".to_string(), "vteilniztwdsoaunpzhztvdddelkndvrwlwctinneujuqkvkvgeeldlkumnqbdzmnklgfortizljnrveczpynqyqqgffyqqunuqp".to_string());
    }

    println!(
        "{}",
        Solution::find_rotate_steps("godding".to_string(), "gd".to_string())
    );
    println!(
        "{}",
        Solution::find_rotate_steps("godding".to_string(), "godding".to_string())
    );
    println!(
        "{}",
        Solution::find_rotate_steps("abcde".to_string(), "ade".to_string())
    );
    println!(
        "{}",
        Solution::find_rotate_steps("y".to_string(), "y".to_string())
    );
    println!("{}", Solution::find_rotate_steps("dccjgocjnscdalkwrlzmsrhnprvvsvqjtqnzirtaasdeldiiokttozjfkwjghhaibtzkoepdvfkhxgmxwtwsrgiryzqljpsntjei".to_string(), "lsdsvjtj".to_string()));
    println!("{}", Solution::find_rotate_steps("zmgddxyivijgullayjbxlmcbirsgpaeqemefpshtehczznvqjaetausoagpkqjtewszqihnzevhbcgpmmtdbyesgxjulyzshnksc".to_string(), "yxebsleaklhkjlx".to_string()));
    println!("{}", Solution::find_rotate_steps("gyzhbpwtmixjdmozuctjctsvcdxtpbtaftuqqirabgmnibnanegqxivxljloodnmmylisompbtccwksvxcdxoixgjibhkmsujdvx".to_string(), "tihsvtysmktbjtjkcvirxbixbx".to_string()));
    println!("{}", Solution::find_rotate_steps("cimrihxbkmotdfudkkrovdydtyamiqdsogsaffnvdqucwfdrtrppmsjsmbekuxzltpktpzcuayjtowoyemlzggktxzkqzosenqgb".to_string(), "dnkvlggcpqslgsbmnyiukfcglsgglvbqvoosetoktrtgezvbukxqsmiefoffbsgdoomhgfmqumsmmdspqckblzcawzupwagsmwii".to_string()));
    println!("{}", Solution::find_rotate_steps("wmgzcptgkwwnezrymhqtxlpbavqpioymjqifbdiwtldfqgcycqlfjxlmyeqtynjpxprzqelgecdqlrbayhmbvneyfoopwlhdkvzc".to_string(), "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa".to_string()));
    println!("{}", Solution::find_rotate_steps("dslgzxnzgjncgeofucxncbjwaehabidlohnklkcgsxugshlhgjsiqjwfxeignrgcfrshxnecfssmqxbcxdckfpodqazgoddkbmsh".to_string(), "fxseifxczzcnhgscdgidkcqshfqbeigceaccszjzxgjghkjdkskjocohphmsdxogxacklrcxcuaxdkscnkdkjajggqxhqkaumgcf".to_string()));
    println!("{}", Solution::find_rotate_steps("qdwcejgziqiulnvnttdkywpbpywduevimnutnzlvfefsntqgdvfrqnfoknvholqozccyhmdjbqizbuvbfafybfuqfdrosqlztazl".to_string(), "vteilniztwdsoaunpzhztvdddelkndvrwlwctinneujuqkvkvgeeldlkumnqbdzmnklgfortizljnrveczpynqyqqgffyqqunuqp".to_string()));
}
